File: game-server/src/main/java/net/dodian/uber/game/runtime/sync/util/LongHashSet.java
package net.dodian.uber.game.runtime.sync.util;

import java.util.Arrays;
import java.util.function.LongConsumer;

public final class LongHashSet {
    private static final float LOAD_FACTOR = 0.5f;

    private long[] keys;
    private int size;
    private int threshold;
    private int mask;

    public LongHashSet(int initialCapacity) {
        int capacity = capacityFor(initialCapacity);
        this.keys = new long[capacity];
        this.mask = capacity - 1;
        this.threshold = Math.max(1, (int) (capacity * LOAD_FACTOR));
    }

    public void clear() {
        Arrays.fill(keys, 0L);
        size = 0;
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean add(long key) {
        long stored = key + 1L;
        if (size + 1 > threshold) {
            resize(keys.length << 1);
        }
        int index = mix(stored) & mask;
        while (true) {
            long existing = keys[index];
            if (existing == 0L) {
                keys[index] = stored;
                size++;
                return true;
            }
            if (existing == stored) {
                return false;
            }
            index = (index + 1) & mask;
        }
    }

    public boolean contains(long key) {
        long stored = key + 1L;
        int index = mix(stored) & mask;
        while (true) {
            long existing = keys[index];
            if (existing == 0L) {
                return false;
            }
            if (existing == stored) {
                return true;
            }
            index = (index + 1) & mask;
        }
    }

    public void forEach(LongConsumer consumer) {
        for (long stored : keys) {
            if (stored != 0L) {
                consumer.accept(stored - 1L);
            }
        }
    }

    private void resize(int newCapacity) {
        long[] oldKeys = keys;
        keys = new long[newCapacity];
        mask = newCapacity - 1;
        threshold = Math.max(1, (int) (newCapacity * LOAD_FACTOR));
        size = 0;
        for (long stored : oldKeys) {
            if (stored != 0L) {
                add(stored - 1L);
            }
        }
    }

    private static int capacityFor(int initialCapacity) {
        int desired = Math.max(2, (int) (initialCapacity / LOAD_FACTOR) + 1);
        int capacity = 1;
        while (capacity < desired) {
            capacity <<= 1;
        }
        return capacity;
    }

    private static int mix(long value) {
        long x = value;
        x ^= (x >>> 33);
        x *= 0xff51afd7ed558ccdL;
        x ^= (x >>> 33);
        x *= 0xc4ceb9fe1a85ec53L;
        x ^= (x >>> 33);
        return (int) x;
    }
}

File: game-server/bin/main/net/dodian/uber/game/engine/sync/util/LongHashSet.kt
package net.dodian.uber.game.engine.sync.util

import java.util.Arrays
import java.util.function.LongConsumer

class LongHashSet(initialCapacity: Int) {
    private var keys: LongArray
    private var size = 0
    private var threshold: Int
    private var mask: Int

    init {
        val capacity = capacityFor(initialCapacity)
        keys = LongArray(capacity)
        mask = capacity - 1
        threshold = maxOf(1, (capacity * LOAD_FACTOR).toInt())
    }

    fun clear() {
        Arrays.fill(keys, 0L)
        size = 0
    }

    fun size(): Int = size

    fun isEmpty(): Boolean = size == 0

    fun add(key: Long): Boolean {
        val stored = key + 1L
        if (size + 1 > threshold) {
            resize(keys.size shl 1)
        }
        var index = mix(stored) and mask
        while (true) {
            val existing = keys[index]
            if (existing == 0L) {
                keys[index] = stored
                size++
                return true
            }
            if (existing == stored) {
                return false
            }
            index = (index + 1) and mask
        }
    }

    fun contains(key: Long): Boolean {
        val stored = key + 1L
        var index = mix(stored) and mask
        while (true) {
            val existing = keys[index]
            if (existing == 0L) {
                return false
            }
            if (existing == stored) {
                return true
            }
            index = (index + 1) and mask
        }
    }

    fun forEach(consumer: LongConsumer) {
        for (stored in keys) {
            if (stored != 0L) {
                consumer.accept(stored - 1L)
            }
        }
    }

    private fun resize(newCapacity: Int) {
        val oldKeys = keys
        keys = LongArray(newCapacity)
        mask = newCapacity - 1
        threshold = maxOf(1, (newCapacity * LOAD_FACTOR).toInt())
        size = 0
        for (stored in oldKeys) {
            if (stored != 0L) {
                add(stored - 1L)
            }
        }
    }

    private companion object {
        private const val LOAD_FACTOR = 0.5f

        private fun capacityFor(initialCapacity: Int): Int {
            val desired = maxOf(2, (initialCapacity / LOAD_FACTOR).toInt() + 1)
            var capacity = 1
            while (capacity < desired) {
                capacity = capacity shl 1
            }
            return capacity
        }

        private fun mix(value: Long): Int {
            var x = value
            x = x xor (x ushr 33)
            x *= -0xae502812aa7333L
            x = x xor (x ushr 33)
            x *= -0x3b314601e57a13adL
            x = x xor (x ushr 33)
            return x.toInt()
        }
    }
}
